capsules/src/mx25r6435f.rs

```diff
@@ -279,17 +284,33 @@ impl<
                             (PAGE_SIZE + 4) as usize,
                         )
                     })
-            })
+            });
+
+        if retval == ReturnCode::SUCCESS {
+            Ok(())
+        } else {
+            Err((retval, self.client_sector.take().unwrap()))
```

Comment:
If `take()` will always return something, then the `map()` or `unwrap_or()` approach will always work, too? Adding `unwrap` may be, in practice, ok, but it is a keyword that signals a potential issue, and isn't code we want people to copy.

Reply:
Using `map` or `unwrap_or` will require the use of an `Option`, since there will no longer be a guarantee of returning a buffer in the event of an error. I think that this would ultimately make the interface much more complicated and just move the `unwrap` into the client's code (since they now have to deal with an `Option` that may or may not contain the buffer).

I also think that not getting a buffer out of a `TakeCell` is a serious error that's hard to recover from, and should never occur in properly written code. While it is possible that such a bug may be introduced, it makes sense to me that this should fail visibly with an `unwrap` rather than failing silently.

Reply:
Ok I agree. What about moving the `self.client_sector.replace(sector);` line into the if block right above `Ok()`? I don't think anything uses client_sector until after both of these functions return.

Reply:
Good catch, that seems like it should be fine as long as the spi driver defers callbacks.